### PR TITLE
hashicorp: deprecate and add caveat

### PR DESCRIPTION
**Pull request by iMichka**
Inform users that we might disable this forumula one day given there will be no more version updates in homebrew-core due to the license change

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

closes #168090

### CODE REVIEW

File: Formula/terraform.rb
Comment (p-linnane) [inline Formula/terraform.rb:48]:
```suggestion
      https://docs.brew.sh/License-Guidelines
```

File: Formula/terraform.rb
Comment (p-linnane) [inline Formula/terraform.rb:50]:
```suggestion
```

Comment (p-linnane) [reply Formula/terraform.rb:50]: Since this will follow the normal deprecation/disable prodecure, I don't think we need this.

File: Formula/terraform.rb
Comment (p-linnane) [inline Formula/terraform.rb:45]:
```suggestion
      The next release will change to a non-open-source license:
```

File: Formula/terraform.rb
Comment (ZhongRuoyu) [inline Formula/terraform.rb:28]:
```suggestion
  deprecate! date: "2023-08-10", because: "will change its license to BSL on the next release"
```

**Comment by Bo98** [comment]
Not sure if it's necessarily worth elaborating on: but future releases prior to 1.6.0 will be acceptable, though it's possible there won't be any such release.

File: Formula/terraform.rb
Comment (carlocab) [inline Formula/terraform.rb:1]: We should add an audit for this. Here's an [example](https://github.com/Homebrew/brew/blob/d30b68715fde642824f9e876b04102fb4f7dfcf6/Library/Homebrew/formula_auditor.rb#L487-L496):
```ruby
    ELASTICSEARCH_KIBANA_RELICENSED_VERSION = "7.11"


    def audit_elasticsearch_kibana
      return if formula.name != "elasticsearch" && formula.name != "kibana"
      return unless @core_tap
      return if formula.version < Version.new(ELASTICSEARCH_KIBANA_RELICENSED_VERSION)


      problem "Elasticsearch and Kibana were relicensed to a non-open-source license from version 7.11. " \
              "They must not be upgraded to version 7.11 or newer."
    end
```

File: Formula/terraform.rb
Comment (MikeMcQuaid) [inline Formula/terraform.rb:44]:
```suggestion
     We will not accept any new Terraform releases in homebrew/core.
```

as it'll already tell you it's deprecated.

File: Formula/terraform.rb
Comment (MikeMcQuaid) [inline Formula/terraform.rb:46]:
```suggestion
        https://www.hashicorp.com/blog/hashicorp-adopts-business-source-license
```

File: Formula/terraform.rb
Comment (MikeMcQuaid) [reply Formula/terraform.rb:48]: (edited to add whitespace)

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for this @iMichka! Looks great so far. A few thoughts.

**Comment by fxcoudert** [comment]
@iMichka gentle ping on that review feedback

File: Formula/terraform.rb
Comment (iMichka) [reply Formula/terraform.rb:1]: Will do in a separate PR

Comment (iMichka) [reply Formula/terraform.rb:1]: See https://github.com/Homebrew/brew/pull/15975

**Comment by iMichka** [comment]
Done. Sorry for the delay

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @iMichka!

File: Formula/t/terraform.rb
Comment (chenrui333) [inline Formula/t/terraform.rb:45]: I think it would only impact 1.6.0, right? if that is the case, we can still accept the new releases with 1.5.x

File: Formula/t/terraform.rb
Comment (chenrui333) [inline Formula/t/terraform.rb:29]: needs to update to 09-12?

**Review by chenrui333 — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @iMichka!

**Review by ZhongRuoyu — COMMENTED** [review]
These unfortunate dependents also need to be deprecated:

```
  atlantis
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  cdktf
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  envconsul
    * Dependency 'consul' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'consul' or deprecate it and all of its dependents.
  fabio
    * Dependency 'consul' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'consul' or deprecate it and all of its dependents.
  infracost
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  stolon
    * Dependency 'consul' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'consul' or deprecate it and all of its dependents.
  terraform-provider-libvirt
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  terraform-rover
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  tfmigrate
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  tfschema
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
  tfupdate
    * Dependency 'terraform' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'terraform' or deprecate it and all of its dependents.
```

**Comment by Bo98** [comment]
Terraform dependents might be resolved when we add OpenTofu, so we could wait a little for that.

Not aware of anything for Consul however.

**Review by p-linnane — APPROVED** [review]
Agreed with @Bo98 on seeing if we can use OpenTofu as a drop in replacement for the Terraform dependents.

**Comment by Bo98** [comment]
I've re-run the tap_syntax to show the failure so we don't try and fail to auto-merge this for a 5th time.

(It previously wasn't showing because we had removed `brew audit` from PR checks for a little while while keeping it for merge checks - but I restored it fully as of yesterday.)

Might be worth splitting consul and terraform to their own PRs. We'll need to make a decision on the former, and for the latter it looks like we'll wait for a little bit until OpenTofu make their first release - we already are blocking version bumps so no rush.

**Comment by ZhongRuoyu** [comment]
We may want to remove the livecheck blocks too -- `vault` 1.15.0 has been released and it's  showing up in the output of `brew bump`.

**Comment by b-reich** [comment]
https://github.com/Homebrew/homebrew-core/pull/149678
I created a formula for the opentofu fork

File: Formula/n/nomad.rb
Comment (bschaatsbergen) [inline Formula/n/nomad.rb:38]: Question; why is 'packer' mentioned in the nomad caveat?

**Comment by thimslugga** [comment]
Perhaps include something to inform those users that want to continue using terraform, etc, that they should use the hashicorp homebrew tap?

https://github.com/hashicorp/homebrew-tap

https://www.hashicorp.com/blog/announcing-hashicorp-homebrew-tap

**Comment by aliscott** [comment]
> These unfortunate dependents also need to be deprecated:

@ZhongRuoyu I'm the lead maintainer of `infracost`. It shouldn't have a dependency on terraform, so we can update the formula to remove that dependency.

**Comment by ZhongRuoyu** [comment]
> It shouldn't have a dependency on terraform, so we can update the formula to remove that dependency.

@aliscott Could you open a PR to update `infracost`'s dependency then? Thanks!

**Comment by MikeMcQuaid** [comment]
> Perhaps include something to inform those users that want to continue using terraform, etc, that they should use the hashicorp homebrew tap?
> 
> https://github.com/hashicorp/homebrew-tap
> 
> https://www.hashicorp.com/blog/announcing-hashicorp-homebrew-tap

We shouldn't link to this directly because:
- we don't want to be seen to specifically recommend a tap we have no influence over
- homebrew-core is for open-source software only and we want to be careful pointing to a repository that is not open source software

**Comment by baggiponte** [comment]
What does this mean for tools like [tfenv](https://github.com/tfutils/tfenv)? Can homebrew be used to install OSS that installs non-OSS software nevertheless?

**Comment by Bo98** [comment]
> What does this mean for tools like [tfenv](https://github.com/tfutils/tfenv)? Can homebrew be used to install OSS that installs non-OSS software nevertheless?

If tfenv itself is able to remain open source that's likely fine.

File: Formula/n/nomad.rb
Comment (iMichka) [reply Formula/n/nomad.rb:38]: That was a copy paste mistake and should be fixed in #150192

**Comment by iMichka** [comment]
I've moved the formulae that could be deprecated right away to another PR: #150192

File: Formula/c/consul.rb
Comment (nikaro) [inline Formula/c/consul.rb:40]:
```suggestion
      We will not accept any new Consul releases in homebrew/core (with the BUSL license).
```

**Comment by aliscott** [comment]
> @aliscott Could you open a PR to update infracost's dependency then? Thanks!

Thanks @ZhongRuoyu! - opened one here: https://github.com/Homebrew/homebrew-core/pull/150286

**Comment by chenrui333** [comment]
I am gonna remove all the terraform formulae dependencies on `terraform` so that we can start the `terraform` formula deprecation process.

**Comment by minamijoyo** [comment]
Hi, all! Thanks for maintaining Homebrew. I am the author of tfmigrate, tfupdate, and tfschema.

Since they have different degrees of dependence on Terraform, I have commented on supplementary information in each pull request.
https://github.com/Homebrew/homebrew-core/pull/150568#issuecomment-1758831866
https://github.com/Homebrew/homebrew-core/pull/150566#issuecomment-1758823693
https://github.com/Homebrew/homebrew-core/pull/150567#issuecomment-1758820948

**Comment by samford** [comment]
I temporarily replaced the `livecheck` blocks for `consul` and `terraform` with `#skip` calls (https://github.com/Homebrew/homebrew-core/pull/151264) while this PR is being worked through. We should remove those `livecheck` blocks here, so livecheck will automatically skip these formulae as deprecated (like the other deprecated Hashicorp formulae).

**Review by chenrui333 — APPROVED** [review]

**Comment by SMillerDev** [comment]
```console
  envconsul
    * Dependency 'consul' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'consul' or deprecate it and all of its dependents.
  fabio
    * Dependency 'consul' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'consul' or deprecate it and all of its dependents.
  stolon
    * Dependency 'consul' is deprecated but has un-deprecated dependents. Either
      un-deprecate 'consul' or deprecate it and all of its dependents.
```
[envconsul](https://formulae.brew.sh/formula/envconsul#default): 20 installs this month
[fabio](https://formulae.brew.sh/formula/fabio#default): 5 installs this month
[stolon](https://formulae.brew.sh/formula/stolon#default): 17 installs this month

I think we can deprecate them, since there is no core-eligible replacement for consul

**Comment by iMichka** [comment]
Rebased and added some comments / updates. Still some work to do here ...

**Comment by chenrui333** [comment]
FYI, HC vault got forked into openbao, https://github.com/openbao/openbao, that would help with vault deprecation.

**Comment by chenrui333** [comment]
> Is there an replacement project for packer?

I dont think so

**Comment by iMichka** [comment]
This should be good now, I took care of all the dependencies.

**Comment by p-linnane** [comment]
Just a nit, but since this has already happened we can probably remove all the future tense language.

**Comment by iMichka** [comment]
> Just a nit, but since this has already happened we can probably remove all the future tense language.

Done; this pull request feels so old now :)

**Review by p-linnane — APPROVED** [review]
Thanks for all your hard work here!

**Comment by p-linnane** [comment]
Might need to open a new PR. The approval isn't registering.

**Review by MikeMcQuaid — APPROVED** [review]

**Review by chenrui333 — APPROVED** [review]

**Pull request merged**
